cookbooks/fb_networkmanager/spec/resource_spec.rb

```diff
@@ -0,0 +1,181 @@
+      expect(File).to receive(:exist?).with(mfile).and_return(true)
+      expect(determine_files('new', { '_migrate_from' => 'Old' })).
+        to eq({ 'config' => cfile, 'migrate' => mfile, 'from' => mfile })
+    end
+
+    it 'uses the target file as source if it exists even ' +
+        'if the migration file exists' do
+      expect(File).to receive(:exist?).with(cfile).and_return(true)
+      allow(File).to receive(:exist?).with(mfile).and_return(true)
+      expect(determine_files('new', { '_migrate_from' => 'Old' })).
+        to eq({ 'config' => cfile, 'migrate' => mfile, 'from' => cfile })
+    end
+
+    it 'uses the target as source if only it exists' do
+      expect(File).to receive(:exist?).with(cfile).and_return(true)
+      allow(File).to receive(:exist?).with(mfile).and_return(false)
+      expect(determine_files('new', { '_migrate_from' => 'Old' })).
+        to eq({ 'config' => cfile, 'migrate' => mfile, 'from' => cfile })
+    end
+
+    it 'uses the target as source if no files exist, but warns' do
+      expect(File).to receive(:exist?).with(cfile).and_return(false)
+      allow(File).to receive(:exist?).with(mfile).and_return(false)
+      expect(Chef::Log).to receive(:warn)
+      expect(determine_files('new', { '_migrate_from' => 'Old' })).
+        to eq({ 'config' => cfile, 'migrate' => mfile, 'from' => cfile })
+    end
+
+    it 'sets source to target if no migration requested; migrate is nil' do
+      allow(File).to receive(:exist?).with(cfile).and_return(true)
+      allow(File).to receive(:exist?).with(mfile).and_return(true)
+      expect(determine_files('new', { 'id' => 'new' })).
+        to eq({ 'config' => cfile, 'migrate' => nil, 'from' => cfile })
+    end
+
+    it 'uses removes _migrate_from key always' do
+      allow(File).to receive(:exist?).with(cfile).and_return(true)
+      allow(File).to receive(:exist?).with(mfile).and_return(true)
+      info = { '_migrate_from' => 'Old', 'id' => 'new' }
+      determine_files('new', info)
+      expect(info).to eq({ 'id' => 'new' })
+    end
+  end
+
+  context '#allowed_connections' do
+    let(:node) { Chef::Node.new }
+    it 'retuns requested connections, normalized' do
+      node.default['fb_networkmanager']['system_connections'] = {
+        'test1' => {},
+        'test2' => {},
+      }
+      expect(allowed_connections(node)).to eq(
+        ['fb_networkmanager_test1', 'fb_networkmanager_test2'],
+      )
+    end
+  end
+
+  context '#generate_config_hashes' do
+    let(:from) do
+      'afile'
+    end
+
+    let(:config) do
+      {
+        'section1' => {
+          'optionA' => 'first_value',
+        },
+        'section2' => {
+          'optionB' => 'second_value',
+        },
+      }
+    end
+
+    let(:from_contents) do
+      <<~EOS
+      [oldsection]
+      key=value
+
+      [section1]
+      existing=stuff
+      optionA=previous_value
+
+      [section2]
+      blue=red
+      EOS
+    end
```

Comment:
Add another test case where there's multi-value fields in the contents. This content:

```
[oldsection]
key=value
foo=bar;baz;

[section1]
existing=stuff
optionA=previous_value
optionB=red;orange;yellow;
```

must generate this hash:

```
{
  'oldsection' => {
    'key' => 'value',
    'foo' => [
      'bar',
      'baz',
    ],
  },
  'section1' => {
    'existing' => 'stuff',
    'optionA' => 'previous_value',
    'optionB' => [
      'red',
      'orange',
      'yellow',
    ],
  },
}
```